imp_act/environments/recorder.py

```diff
@@ -0,0 +1,87 @@
+"""
+The Recorder class is used to record the interactions of the agent with 
+the environment. It records the observations, actions, rewards, done flags,
+in a "tape". It is a wrapper around the environment.
+
+The tape is a dict for 3 reasons:
+    - Each episode can have varying keys, 
+        for example, if there is an evaluation at that episode etc.
+    - Easy manipulation using pandas,
+    - Easy conversion to csv for storage where it is human readable
+
+"""
+
+import pandas as pd
+
+
+class Recorder:
+    def __init__(self, env):
+        self.env = env
+        self.init_recorder()
+
+    def init_recorder(self):
+        self.time_step = 0
+        self.episode = -1  # because we increment it at the beginning of reset
+
+        self.tape = {}
+
+    def reset(self):
+        self.time_step = 0
+        self.episode += 1
+
+        obs = self.env.reset()
+
+        # Record
+        self.tape[self.episode] = {
+            # lists with 'max_timesteps+1' elements
+            "time_step": [self.time_step],
+            "edge_states": [self.env._get_states()],
+            "edge_observations": [obs["edge_observations"]],
+            "edge_deterioration_rates": [obs["edge_deterioration_rates"]],
+            "edge_beliefs": [obs["edge_beliefs"]],
+            # lists with 'max_timesteps' elements
+            "action": [],
+            "reward": [],
+            "done": [],
+            "total_travel_time": [],
+            "travel_times": [],
+            "volumes": [],
+            "reward_elements": [],
+            "actions_taken": [],
+        }
+
+        return obs
+
+    def step(self, action):
+        obs, reward, done, info = self.env.step(action)
+
+        # Record
+        self.record("time_step", self.time_step)
```

Comment:
I tried something like that in the first commit [here](https://github.com/AI-for-Infrastructure-Management/imp-act/commit/6e7e3c90dfc49a30bb872435eae6461cb6d4c79f), but decided to manually add keys because,
- But doing it in this way made accessing the objects inside the 'obs' variable becomes quite cluttered. 
- Some keys such as 'edge_observations' have max_timesteps+1 entries, but 'edge_observations' are not returned by reset, but rather only in 'info', so to access it here, I have to use 'self.env._get_states()'.
 
Overall, I couldn't figure out how to loop over the keys in a "clean" way.

Reply:
![grafik](https://github.com/user-attachments/assets/e445781d-bf91-4ebd-a775-412583c6a1b6)

What do you think about something like this?
This way the only naming dependency is on the edge_states and the rest is updated automatically should we change the observations.

I can push the changes if you agree.

Reply:
yes, I think this is much better!

One last thing, we included the "adjacency_matrix" in obs for the case when graph connectivity changes, right? If that feature is not implemented right now, does it makes sense to remove it from the obs dict? This way we can avoid the "record_exclude_keys". Because in the case when connectivity changes, it must recorded?

Reply:
Cool, thanks Prateek :) 
I think I would leave it as part of the observations, it might already be usefull to the agent to determine the redundancies. And for the challenge the agents should only rely on what is passed using the observation dict, (if the object can be accessed the agent could change rewards, see states etc. ) so it would be best to keep it in.